File: src/2517_Maximum_Tastiness_of_Candy_Basket.py
# https://leetcode.cn/problems/maximum-tastiness-of-candy-basket/

class Solution:
    def maximumTastiness(self, price: List[int], k: int) -> int:
        price.sort()
        left, right = 0, price[-1] - price[0]
        while left < right:
            mid = (left + right + 1) // 2
            if self.check(price, k, mid):
                left = mid
            else:
                right = mid - 1
        return left

    def check(self, price: List[int], k: int, tastiness: int) -> bool:
        prev = -inf
        cnt = 0
        for p in price:
            if p - prev >= tastiness:
                cnt += 1
                prev = p
        return cnt >= k



class Solution:
    def maximumTastiness(self, price: List[int], k: int) -> int:
        price.sort()

        def check(limit: int) -> bool:
            x1 = price[0] + limit
            count = 1
            for p in price:
                if p >= x1:
                    count += 1
                    x1 = p + limit
            return count >= k

        left = 1
        right = (price[-1] - price[0]) // (k - 1)
        while left <= right:
            mid = (left + right) // 2
            if check(mid):
                left = mid + 1
            else:
                right = mid - 1

        return right

File: src/2517_Maximum_Tastiness_of_Candy_Basket.rs
// 2517_Maximum_Tastiness_of_Candy_Basket
// https://leetcode.cn/problems/maximum-tastiness-of-candy-basket/
impl Solution {
    pub fn maximum_tastiness(price: Vec<i32>, k: i32) -> i32 {
        let mut price = price;
        let len = price.len();
        price.sort_unstable();
        let mut left = 0;
        let mut right = price[len-1] - price[0];
        while left < right {
            let mid = (left + right + 1) >> 1;
            if check(&price, k, mid) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }
        return left;

    }
}

pub fn check(price: &Vec<i32>, k: i32, tastiness: i32) -> bool {
    let mut prev = price[0];
    let mut cnt = 1;
    for p in price.iter().skip(1) {
        if *p - prev >= tastiness {
            cnt += 1;
            prev = *p;
        }
    }
    return cnt >= k;
}




impl Solution {
    pub fn maximum_tastiness(mut price: Vec<i32>, k: i32) -> i32 {
        let check = |price: &[i32], diff: i32| -> bool {
            let mut cnt = 1;
            let mut lst = price[0];
            for &x in price.iter().skip(1) {
                if x >= lst + diff {
                    cnt += 1;
                    lst = x;
                }
            }
            cnt >= k
        };

        price.sort_unstable();
        let mut i = 0;
        let mut j = (*price.last().unwrap() - price[0]) / (k - 1) + 1;
        while i < j {
            let m = (i + j + 1) >> 1;
            if check(&price, m) {
                i = m;
            } else {
                j = m - 1;
            }
        }
        i
    }
}